src/cell.cpp

```diff
@@ -635,6 +635,93 @@ void DAGCell::to_hdf5(hid_t group_id) const { return; }
 
 #endif
 
+//==============================================================================
+// UniversePartitioner implementation
+//==============================================================================
+
+UniversePartitioner::UniversePartitioner(const Universe& univ)
+{
+  // Find all of the z-planes in this universe.  Add them to the surfs_ member.
+  // Use surf_set for O(1) searches that ensure surfs_ entries are not repeated.
+  std::unordered_set<int32_t> surf_set;
+  for (auto i_cell : univ.cells_) {
+    for (auto token : model::cells[i_cell]->rpn_) {
+      if (token < OP_UNION) {
+        auto i_surf = std::abs(token) - 1;
+        const auto* surf = model::surfaces[i_surf].get();
+        if (const auto* zplane = dynamic_cast<const SurfaceZPlane*>(surf)) {
+          if (surf_set.find(i_surf) == surf_set.end()) {
+            surf_set.insert(i_surf);
+            surfs_.push_back(i_surf);
+          }
+        }
+      }
+    }
+  }
+
+  // Define a functor for comparing z-planes by their location, and use it to
+  // sort the surfs_ member.
+  struct {
+    bool operator()(int32_t i_surf, int32_t j_surf) const
+    {
+      const auto* surf = model::surfaces[i_surf].get();
+      const auto* zplane = dynamic_cast<const SurfaceZPlane*>(surf);
+      double zi = zplane->z0_;
+      surf = model::surfaces[j_surf].get();
+      zplane = dynamic_cast<const SurfaceZPlane*>(surf);
+      double zj = zplane->z0_;
+      return zi < zj;
+    }
+  } compare_surfs;
+  std::sort(surfs_.begin(), surfs_.end(), compare_surfs);
+
+  // Populate the partition lists.
+  partitions_.resize(surfs_.size() + 1);
+  for (auto i_cell : univ.cells_) {
+    // Find the tokens for bounding z-planes.
+    int32_t min_token = 0, max_token = 0;
+    double min_z, max_z;
+    for (auto token : model::cells[i_cell]->rpn_) {
+      if (token < OP_UNION) {
+        const auto* surf = model::surfaces[std::abs(token) - 1].get();
+        if (const auto* zplane = dynamic_cast<const SurfaceZPlane*>(surf)) {
+          if (min_token == 0 || zplane->z0_ < min_z) {
+            min_token = token;
+            min_z = zplane->z0_;
+          }
+          if (max_token == 0 || zplane->z0_ > max_z) {
+            max_token = token;
+            max_z = zplane->z0_;
+          }
+        }
+      }
+    }
+
+    // If there are no bounding z-planes, add this cell to all partitions.
+    if (min_token == 0) {
+      for (auto& p : partitions_) p.push_back(i_cell);
+      continue;
+    }
+
+    // Iterate over partitions, and add this cell to each appropriate one.
+    // Since surfs_ is sorted, we know this cell belongs to every partition
+    // between min_token and max_token.
+    bool in_partition = min_token < 0;
```

Comment:
What happens if someone has defined a universe with lots of z-planes, but they don't define any cell that has `-min_zplane` (maybe because it gets cut off by virtue of being filled in another cell)? In that case, `min_token` would be positive, right?

Reply:
Yeah, we actually already have a case in the test suite that triggers this kind of weird condition.  The void regression test has enough z-planes to trigger the partitioning and it has a bunch of cells that are bound on one side by z-plane and by a sphere on the other.  In this case you can have a `min_token` that is negative.  (In this case, `min_token` is also equal to `max_token`.  So the code here assumes that the cell is present in the bottom partition and then every partition up to the `max_token == min_token` z-plane.)

In the normal case that we have a cell sandwiched between two z-planes then `min_token` is positive and `max_token` is negative.  I just now realized how counter-intuitive that is so maybe I should rename them to `lower_token` and `upper_token`

Reply:
I'm not sure I fully understand yet, but it sounds like you've accounted for this case. Maybe after the renaming things will be more obvious to me :)